marklogic-client-api/src/main/java/com/marklogic/client/datamovement/JSONSplitter.java

```diff
@@ -0,0 +1,550 @@
+                                    break;
+
+                                default:
+                                    throw new IllegalStateException("Unknown state");
+                            }
+
+                            break;
+
+                        case END_ARRAY:
+                            visitor.endArray(key.peek());
+                            break;
+                    }
+                }
+
+                return null;
+
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to traverse document", e);
+            }
+        }
+
+    }
+
+    private static class HandleSpliterator<T extends AbstractWriteHandle> extends JSONSpliterator<T, T> {
+
+        HandleSpliterator(JSONSplitter<T> splitter, JsonParser jsonParser) {
+            super(splitter, jsonParser);
+        }
+
+        @Override
+        public boolean tryAdvance(Consumer<? super T> action) {
+            T handle = (T) getNextHandle();
+            if (handle == null) {
+                return false;
+            }
+
+            action.accept(handle);
+            getSplitter().count++;
+
+            return true;
+        }
+    }
+
+    private static class DocumentWriteOperationSpliterator<T extends AbstractWriteHandle> extends JSONSpliterator<DocumentWriteOperation, T> {
+
+        DocumentWriteOperationSpliterator(JSONSplitter<T> splitter, JsonParser jsonParser) {
+            super(splitter, jsonParser);
+        }
+
+        @Override
+        public boolean tryAdvance(Consumer<? super DocumentWriteOperation> action) {
+            T handle = (T) getNextHandle();
+            if (handle == null) {
+                return false;
+            }
+
+            DocumentWriteOperation documentWriteOperation = getSplitter().getVisitor().makeDocumentWriteOperation(handle);
+
+            action.accept(documentWriteOperation);
+            getSplitter().count++;
+
+            return true;
+        }
+    }
+
+    private static class JsonContainerParser extends JsonParserDelegate {
+
+        JsonContainerParser(JsonParser jsonParser) {
+            super(jsonParser);
+        }
+
+        private int depth = 1;
+
+        @Override
+        public boolean isClosed() {
+            return (depth > 0);
+        }
+
+        private void maintainDepth(JsonToken currentToken) {
+            if (currentToken == JsonToken.START_ARRAY || currentToken == JsonToken.START_OBJECT) {
+                depth++;
+            }
+
+            if (currentToken == JsonToken.END_ARRAY || currentToken == JsonToken.END_OBJECT) {
+                depth--;
+            }
+        }
+
+        private void maintainDepth() {
+            maintainDepth(super.currentToken());
+        }
+
+        @Override
+        public JsonToken nextToken() throws IOException {
+            if (depth == 0) {
+                throw new IllegalStateException("The JSON branch is closed");
+            }
+
+            JsonToken next = super.nextToken();
```

Comment:
Would it be simpler to call the maintainDepth() method that takes no arguments?  Or even to combine the maintainDepth() methods as a method that takes no arguments and has line 490 as its first statement?

Reply:
I changed maintainDepth() as below

```
private void maintainDepth() throws IOException {
            JsonToken next = super.nextToken();
            maintainDepth(next);
        }
```

And changed blow function to use maintainDepth() alone like below
```
@Override
        public Boolean nextBooleanValue() throws IOException {
            if (depth == 0) {
                return null;
            }

            maintainDepth();
            return super.getBooleanValue();
        }
```

I'm also wondering is the logic for isClosed() wrong? Should we return (depth <= 0) instead of (depth > 0). Then when depth is 0, it returns true.

Reply:
That makes sense.  Good catch.